File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # Binary Search and Merging;

        n = len(intervals)
        def binary_search(interval, intervals):
            target = interval[0]

            left = 0
            right = n - 1
            insertion_idx = -1

            while left <= right:
                mid = left + (right - left) // 2

                if intervals[mid][0] > target:
                    right = mid - 1
                else:
                    insertion_idx = mid
                    left = mid + 1
            
            intervals = intervals[:insertion_idx + 1] + [newInterval] + intervals[insertion_idx + 1:]
            return intervals

        intervals = binary_search(newInterval, intervals)

        merged_intervals = []
        prev_interval = intervals[0]

        # print(intervals)

        for idx in range(1, n + 1):
            curr_interval = intervals[idx]

            if prev_interval[1] >= curr_interval[0]:
                # print(prev_interval)
                prev_interval[1] = max(prev_interval[1], curr_interval[1])
            else:
                merged_intervals.append(prev_interval)
                prev_interval = curr_interval
        
        # print(prev_interval)
        merged_intervals.append(prev_interval)

        return merged_intervals        

File: 0057-insert-interval/0057-insert-interval.java
class Solution {
    private static int lowerBound(int[][] intervals, int target){
        int n = intervals.length;

        int left = 0;
        int right = n - 1;
        int insertionIdx = -1;

        while(left <= right){
            int mid = left + (right - left) / 2;

            if(intervals[mid][0] > target){
                right = mid - 1;
            }
            else{
                insertionIdx = mid;
                left = mid + 1;
            }
        }

        return insertionIdx;
    }

    private static int[][] mergedIntervals(int[][] intervals){
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        ArrayList<int[]> mergedList = new ArrayList<>();
		
		int[] prevInterval = intervals[0];
		
		for(int idx = 1; idx < intervals.length; idx++){
			int[] currInterval = intervals[idx];
			
			if(prevInterval[1] >= currInterval[0]){
				prevInterval[1] = Math.max(prevInterval[1], currInterval[1]);
			}
			else{
				mergedList.add(prevInterval);
				prevInterval = currInterval;
			}
		}
		
		mergedList.add(prevInterval);
		
		int[][] mergedIntervals = new int[mergedList.size()][];
		
		mergedList.toArray(mergedIntervals);
		return mergedIntervals;
    }

    public int[][] insert(int[][] intervals, int[] newInterval) {
        int insertionIdx = lowerBound(intervals, newInterval[0]);

        int[][] newIntervals = new int[intervals.length + 1][2];

        for(int idx = 0; idx <= insertionIdx; idx++){
            newIntervals[idx] = intervals[idx];
        }

        newIntervals[insertionIdx + 1] = newInterval;

        for(int idx = insertionIdx + 1; idx < intervals.length; idx++){
            newIntervals[idx + 1] = intervals[idx];
        }

        int[][] mergedIntervals = mergedIntervals(newIntervals);
        return mergedIntervals;
    }
}